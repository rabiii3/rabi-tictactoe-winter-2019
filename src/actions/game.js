import { GAME_STARTED, getNewBoard, getNextPlayer, updateBoard, GAME_OVER, isBoardFull, randomPlayer } from '../game';
import { requestComputerPlay, hasAWinner } from '../request';
import { loadFruits } from './fruit';

export const STARTED = 'STARTED';
export const PLAYED = 'PLAYED';
export const PLAYER_CHANGED = 'PLAYER_CHANGED';
export const ENDED = 'ENDED';

export const startGameAction = () => (dispatch, getState) => {
  const {
    game: { player, computer },
  } = getState();
  dispatch({
    type: STARTED,
    payloads: {
      board: getNewBoard(),
      status: GAME_STARTED,
      currentPlayer: randomPlayer(player, computer),
    },
  });
  const {
    game: { currentPlayer },
  } = getState();
  if (currentPlayer.isComputer) dispatch(computerPlay());
};

export const playerPlay = index => (dispatch, getState) => {
  const {
    game: { board, currentPlayer },
  } = getState();
  const newBoard = updateBoard(board, index, currentPlayer);
  dispatch(manageGame(newBoard));
};

const computerPlay = () => (dispatch, getState) => {
  const {
    game: { board, currentPlayer, player, computer },
  } = getState();
  return requestComputerPlay({ board, computer: computer.piece, player: player.piece }).then(({ move }) => {
    const newBoard = updateBoard(board, move, currentPlayer);
    setTimeout(() => dispatch(manageGame(newBoard)), 1000);
  });
};

export const manageGame = newBoard => (dispatch, getState) => {
  dispatch({
    type: PLAYED,
    payloads: {
      board: newBoard,
    },
  });

  const {
    game: { board, currentPlayer, player, computer },
  } = getState();
  hasAWinner({ board }).then(({ winner }) => {
    if (winner || isBoardFull(board)) {
      dispatch(loadFruits()).then(() =>
        dispatch({
          type: ENDED,
          payloads: {
            result: winner,
            player,
            computer,
            status: GAME_OVER,
          },
        }),
      );
    } else {
      dispatch({
        type: PLAYER_CHANGED,
        payloads: {
          currentPlayer: getNextPlayer(currentPlayer, player, computer),
        },
      });

      const {
        game: { currentPlayer: cp },
      } = getState();
      console.log('======currennt player :', cp);
      if (cp.isComputer) dispatch(computerPlay());
    }
  });
};
