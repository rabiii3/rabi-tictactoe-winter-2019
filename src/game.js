import { indexOf, update, propOr } from 'ramda';

export const X = 'x';

export const O = 'o';

export const GAME_OVER = 'GAME_OVER';

export const GAME_STARTED = 'GAME_STARTED';

export const isBoardFull = board => indexOf(null, board) < 0;

export const getStatus = board => (isBoardFull(board) ? GAME_OVER : GAME_STARTED);

export const getNewComputerBoard = (board, computer) => update(indexOf(null, board), computer.piece, board);

export const getNewBoard = () => [null, null, null, null, null, null, null, null, null];

export const getFirstAvailableCell = board => indexOf(null, board);

export const updateBoard = (board, index, currentPlayer) =>
  isBoardFull(board) ? board : update(index, currentPlayer.piece, board);

export const getNextPlayer = (currentPlayer, player, computer) => (currentPlayer.isComputer ? player : computer);

const random = () => Math.round(Math.random());

export const randomPlayer = (player, computer) => {
  const players = [player, computer];
  return players[random()];
};

export const updateFruitBoard = (board, index, icon) => update(index, icon, board);

export const isFruit = piece => (piece != null && propOr(false, 'icon', piece) ? true : false);
