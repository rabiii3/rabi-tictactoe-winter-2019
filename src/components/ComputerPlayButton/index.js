import React from 'react';
import PropTypes from 'prop-types';
import { Button } from '@material-ui/core';
import { GAME_OVER } from '../../game';

const isStatusOver = status => status === GAME_OVER;

const ComputerPlayButton = ({ status, onClick, children, currentPlayer }) => (
  <Button onClick={onClick} disabled={isStatusOver(status) || !currentPlayer.isComputer}>
    {children}
  </Button>
);

ComputerPlayButton.propTypes = {
  status: PropTypes.node.isRequired,
  onClick: PropTypes.func.isRequired,
  currentPlayer: PropTypes.object.isRequired,
  children: PropTypes.node,
};

export default ComputerPlayButton;
