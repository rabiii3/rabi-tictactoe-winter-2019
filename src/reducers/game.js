import { STARTED, PLAYED, PLAYER_CHANGED, ENDED } from '../actions/game';
import { CELL_FRUIT_LOADED } from '../actions/fruit';
import { updateFruitBoard } from '../game';

export default (state = {}, action) => {
  var i = 0;
  switch (action.type) {
    case STARTED:
      return { ...state, ...action.payloads };
    case PLAYED:
      return { ...state, ...action.payloads };
    case PLAYER_CHANGED:
      return { ...state, ...action.payloads };
    case ENDED: {
      const { status } = action.payloads;
      return { ...state, status };
    }
    case CELL_FRUIT_LOADED: {
      const { board } = state;
      const { index, fruit } = action;
      return { ...state, board: updateFruitBoard(board, index, fruit) };
    }
    default:
      return state;
  }
};
