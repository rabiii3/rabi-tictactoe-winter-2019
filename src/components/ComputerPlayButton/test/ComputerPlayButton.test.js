import React from 'react';
import { shallow } from 'enzyme';
import ComputerPlayButton from '..';
import { GAME_OVER, X } from '../../../game';

describe('ComputerPlayButton', () => {
  it('should render', () => {
    const props = {
      onClick: jest.fn,
      currentPlayer: { name: 'test', piece: X },
      status: GAME_OVER,
    };
    const root = <ComputerPlayButton {...props} />;
    const wrapped = shallow(root);
    expect(wrapped).toMatchSnapshot();
  });
});
