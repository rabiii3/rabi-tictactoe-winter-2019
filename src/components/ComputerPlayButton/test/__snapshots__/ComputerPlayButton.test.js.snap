// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComputerPlayButton should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ComputerPlayButton
    currentPlayer={
      Object {
        "name": "test",
        "piece": "x",
      }
    }
    onClick={[Function]}
    status="GAME_OVER"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": undefined,
      "disabled": true,
      "onClick": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": undefined,
        "disabled": true,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
